fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/properties/RecordTypesProperty.java

```diff
@@ -81,6 +82,10 @@ public boolean shouldVisit(@Nonnull ExpressionRef<? extends PlannerExpression> r
                     .map(RecordType::getName).collect(Collectors.toSet());
         } else if (relationalExpression instanceof TypeFilterExpression) {
             return Sets.filter(childResults.get(0), ((TypeFilterExpression)relationalExpression).getRecordTypes()::contains);
+        } else if (relationalExpression instanceof LogicalIndexScanExpression) {
+            Index index = context.getIndexByName(((LogicalIndexScanExpression)relationalExpression).getIndexName());
+            return context.getMetaData().recordTypesForIndex(index).stream()
+                    .map(RecordType::getName).collect(Collectors.toSet());
```

Comment:
This might be less than the full set if the expression has an equality predicate on the record type key, but maybe that's fine. I suppose the same is true for a `RecordQueryPlanWithIndex`.